File: app-mvp/src/main/java/xuk/imaginary/gui/gank/meizhi/MeizhiFragment.java
package xuk.imaginary.gui.gank.meizhi;

import android.app.ActivityOptions;
import android.content.Intent;
import android.os.Bundle;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import androidx.annotation.Nullable;
import androidx.core.app.ActivityCompat;
import androidx.fragment.app.Fragment;
import androidx.recyclerview.widget.RecyclerView;
import androidx.recyclerview.widget.StaggeredGridLayoutManager;

import com.chad.library.adapter.base.BaseQuickAdapter;
import com.cjj.MaterialRefreshLayout;
import com.cjj.MaterialRefreshListener;

import java.util.ArrayList;
import java.util.List;

import xuk.imaginary.R;
import xuk.imaginary.data.BaseBean;
import xuk.imaginary.data.source.GankRepository;
import xuk.imaginary.gui.gank.show.GankActivity;
import xuk.imaginary.util.DateUtils;
import xuk.imaginary.util.Utils;

public class MeizhiFragment extends Fragment implements MeizhiContract.View {
    private static final String TAG = "MeizhiFragment";

    public static MeizhiFragment getInstance() {
        return new MeizhiFragment();
    }

    private MeizhiContract.Presenter mPresenter;
    private final MeiZhiAdapter meiZhiAdapter = new MeiZhiAdapter(null);

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        new MeizhiPresenter(GankRepository.getInstance(), this);
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        View v = inflater.inflate(R.layout.fragment_meizhi, container, false);
        RecyclerView recyclerView = v.findViewById(R.id.recycler_view);
        MaterialRefreshLayout materialRefresh = v.findViewById(R.id.material_refresh);
        Log.d(TAG, "onCreateView");

        final StaggeredGridLayoutManager layoutManager = new StaggeredGridLayoutManager(2, StaggeredGridLayoutManager.VERTICAL);
        recyclerView.setLayoutManager(layoutManager);
        recyclerView.setAdapter(meiZhiAdapter);
        meiZhiAdapter.setOnItemClickListener((adapter, view, position) -> {
            Log.d(TAG, "position:" + position);
            BaseBean fl = (BaseBean) adapter.getData().get(position);
            Log.d("MeizhiFragment", fl.toString());
            String dateString = DateUtils.getDate(fl.getPublishedAt());
            Intent intent = new Intent(getActivity(), GankActivity.class);
            intent.putExtra("date", dateString);
            intent.putExtra("image", fl.getUrl());
            ActivityOptions options = ActivityOptions.makeSceneTransitionAnimation(getActivity(), view, "shareObject");
            ActivityCompat.startActivity(getActivity(), intent, options.toBundle());
        });
        meiZhiAdapter.openLoadAnimation(BaseQuickAdapter.ALPHAIN);
        meiZhiAdapter.isFirstOnly(true);

        materialRefresh.setMaterialRefreshListener(new MaterialRefreshListener() {
            @Override
            public void onRefresh(MaterialRefreshLayout materialRefreshLayout) {
                mPresenter.getList(true);
                materialRefreshLayout.postDelayed(materialRefreshLayout::finishRefresh, 1000);
            }

            @Override
            public void onRefreshLoadMore(MaterialRefreshLayout materialRefreshLayout) {
                super.onRefreshLoadMore(materialRefreshLayout);
                mPresenter.getList(false);
                materialRefreshLayout.postDelayed(materialRefreshLayout::finishRefreshLoadMore, 1000);
            }
        });

        mPresenter.subscribe();
        return v;
    }

    @Override
    public void setPresenter(MeizhiContract.Presenter presenter) {
        mPresenter = Utils.checkNotNull(presenter);
    }

    @Override
    public void addList(boolean isRefresh, List<BaseBean> meiZhis) {
        if (isRefresh) {
            meiZhiAdapter.replaceData(new ArrayList<>());
        }
//        Log.d(TAG, meiZhis.toString());
        meiZhiAdapter.addData(meiZhis);
    }

    @Override
    public void onDestroyView() {
        super.onDestroyView();
        mPresenter.unsubscribe();
    }
}

File: app-mvp-kotlin/src/main/java/xuk/imaginary/gui/gank/meizhi/MeizhiFragment.kt
package xuk.imaginary.gui.gank.meizhi

import android.app.ActivityOptions
import android.content.Intent
import android.os.Bundle
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.core.app.ActivityCompat
import androidx.fragment.app.Fragment
import androidx.recyclerview.widget.RecyclerView
import androidx.recyclerview.widget.StaggeredGridLayoutManager
import com.chad.library.adapter.base.BaseQuickAdapter
import com.cjj.MaterialRefreshLayout
import com.cjj.MaterialRefreshListener
import xuk.imaginary.R
import xuk.imaginary.data.BaseBean
import xuk.imaginary.data.source.GankRepository
import xuk.imaginary.gui.gank.show.GankActivity
import xuk.imaginary.util.DateUtils
import java.util.*

/**
 * A simple [Fragment] subclass.
 */
class MeizhiFragment : Fragment(), MeizhiContract.View {

  override lateinit var presenter: MeizhiContract.Presenter
  private lateinit var recyclerView: RecyclerView
  private lateinit var materialRefresh: MaterialRefreshLayout

  private val meiZhiAdapter = MeiZhiAdapter(null)

  override fun onCreate(savedInstanceState: Bundle?) {
    super.onCreate(savedInstanceState)
    MeizhiPresenter(GankRepository.instance, this)
  }

  override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
    val v = inflater.inflate(R.layout.fragment_meizhi, container, false)
    Log.d(TAG, "onCreateView")

    with(v) {
      recyclerView = findViewById(R.id.recycler_view)
      materialRefresh = findViewById(R.id.material_refresh)
    }

    val layoutManager = StaggeredGridLayoutManager(2, StaggeredGridLayoutManager.VERTICAL)
    recyclerView.layoutManager = layoutManager
    recyclerView.adapter = meiZhiAdapter
    meiZhiAdapter.setOnItemClickListener { adapter, view, position ->
      Log.d(TAG, "position:$position")
      val fl = adapter.data[position] as BaseBean
      Log.d(TAG, fl.toString())
      val dateString = DateUtils.getDate(fl.publishedAt)
      val intent = Intent(activity, GankActivity::class.java)
      intent.putExtra("date", dateString)
      intent.putExtra("image", fl.url)
      val options = ActivityOptions.makeSceneTransitionAnimation(activity, view, "shareObject")
      ActivityCompat.startActivity(requireActivity(), intent, options.toBundle())
    }
    meiZhiAdapter.openLoadAnimation(BaseQuickAdapter.ALPHAIN)
    meiZhiAdapter.isFirstOnly(true)

    materialRefresh.setMaterialRefreshListener(object : MaterialRefreshListener() {
      override fun onRefresh(materialRefreshLayout: MaterialRefreshLayout) {
        presenter.getList(true)
        materialRefreshLayout.postDelayed({ materialRefreshLayout.finishRefresh() }, 1000)
      }

      override fun onRefreshLoadMore(materialRefreshLayout: MaterialRefreshLayout) {
        super.onRefreshLoadMore(materialRefreshLayout)
        presenter.getList(false)
        materialRefreshLayout.postDelayed({ materialRefreshLayout.finishRefreshLoadMore() }, 1000)
      }
    })

    presenter.start()
    return v
  }

  override fun addList(isRefresh: Boolean, meiZhis: List<BaseBean>) {
    if (isRefresh)
      meiZhiAdapter.replaceData(ArrayList())
//        Log.d(TAG, meiZhis.toString());
    meiZhiAdapter.addData(meiZhis)
  }

  companion object {
    private const val TAG = "MeizhiFragment"

    val instance: MeizhiFragment
      get() = MeizhiFragment()
  }
}
